payroll/models/hr_payslip.py

```diff
@@ -0,0 +1,581 @@
+                    localdict[rule.code] = tot_rule
+                    rules_dict[rule.code] = rule
+                    # sum the amount for its salary category
+                    localdict = _sum_salary_rule_category(
+                        localdict, rule.category_id, tot_rule - previous_amount
+                    )
+                    # create/overwrite the rule in the temporary results
+                    result_dict[key] = {
+                        "salary_rule_id": rule.id,
+                        "contract_id": contract.id,
+                        "name": rule.name,
+                        "code": rule.code,
+                        "category_id": rule.category_id.id,
+                        "sequence": rule.sequence,
+                        "appears_on_payslip": rule.appears_on_payslip,
+                        "condition_select": rule.condition_select,
+                        "condition_python": rule.condition_python,
+                        "condition_range": rule.condition_range,
+                        "condition_range_min": rule.condition_range_min,
+                        "condition_range_max": rule.condition_range_max,
+                        "amount_select": rule.amount_select,
+                        "amount_fix": rule.amount_fix,
+                        "amount_python_compute": rule.amount_python_compute,
+                        "amount_percentage": rule.amount_percentage,
+                        "amount_percentage_base": rule.amount_percentage_base,
+                        "register_id": rule.register_id.id,
+                        "amount": amount,
+                        "employee_id": contract.employee_id.id,
+                        "quantity": qty,
+                        "rate": rate,
+                    }
+                else:
+                    # blacklist this rule and its children
+                    blacklist += [id for id, seq in rule._recursive_search_of_rules()]
+
+        return list(result_dict.values())
+
+    # YTI TODO To rename. This method is not really an onchange, as it is not
+    # in any view employee_id and contract_id could be browse records
+    def onchange_employee_id(
+        self, date_from, date_to, employee_id=False, contract_id=False
+    ):
+        # defaults
+        res = {
+            "value": {
+                "line_ids": [],
+                # delete old input lines
+                "input_line_ids": [(2, x) for x in self.input_line_ids.ids],
+                # delete old worked days lines
+                "worked_days_line_ids": [(2, x) for x in self.worked_days_line_ids.ids],
+                # 'details_by_salary_head':[], TODO put me back
+                "name": "",
+                "contract_id": False,
+                "struct_id": False,
+            }
+        }
+        if (not employee_id) or (not date_from) or (not date_to):
+            return res
+        ttyme = datetime.combine(fields.Date.from_string(date_from), time.min)
```

Comment:
```suggestion
        ttyme = datetime.combine(date_from, time.min)
```
version 12+ Odoo returns always native Python objects

ref : https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-12.0#tasks-to-do-in-the-migration